dask_ml/model_selection/_hyperband.py

```diff
@@ -0,0 +1,665 @@
+    best_score : float
+        The best score, updated live as soon as scores are received.
+        This value is available even if
+        :func:`~dask_ml.model_selection.HyperbandCV` does not
+        complete. Note that this value may be higher than ``best_score_``
+        because intermediate values are considered before the model is
+        finished training.
+    best_params : dict
+        The parameters corresponding to ``best_score``, updated live as
+        soon as scores are received.  This value is available even if
+        :func:`~dask_ml.model_selection.HyperbandCV` does not complete.
+
+    Notes
+    -----
+    Hyperband is state of the art via an adaptive scheme. Hyperband
+    only spends time on high-performing models, because our goal is to find
+    the highest performing model. This means that it stops training models
+    that perform poorly.
+
+    The asynchronous variant [2]_ is controlled by the ``asynchronous``
+    keyword and suited for highly parallel architectures. Architectures with
+    little parallism (i.e., few workers) will benefit from
+    ``asynchronous=False``.
+
+    If dask arrays are passed to
+    :func:`~dask_ml.model_selection.HyperbandCV.fit`, the estimator's
+    ``partial_fit`` method is called over each chunk of the array.
+    Determining how to chunk the array is not straightforward because
+    the number of ``partial_fit`` calls to reach converegence will depend on
+    the chunk size.
+
+    We recommend using time limits to determine how to chunk the
+    input arrays and how to choose ``max_iter``. We recommend
+    using :func:`~dask_ml.model_selection.HyperbandCV.fit_metadata` to help
+    choosing. For example, let's say ``num_epochs=5`` epochs or
+    passes over ``X`` are required for convergence and we are
+    constrained to 1500 partial_fit calls (by a deadline or something).
+    Then we recommend setting ``max_iter=81`` and the chunks of ``X``
+    to have a fractional size of ``num_epochs / max_iter = 0.0617...``
+    of the total size.
+
+    References
+    ----------
+    .. [1] "Hyperband: A novel bandit-based approach to hyperparameter
+           optimization", 2016 by L. Li, K. Jamieson, G. DeSalvo, A.
+           Rostamizadeh, and A. Talwalkar.  https://arxiv.org/abs/1603.06560
+    .. [2] "Massively Parallel Hyperparameter Tuning", 2018 by L. Li, K.
+            Jamieson, A. Rostamizadeh, K. Gonina, M. Hardt, B. Recht, A.
+            Talwalkar.  https://openreview.net/forum?id=S1Y7OOlRZ
+
+    """
+
+    def __init__(
+        self,
+        model,
+        params,
+        max_iter=81,
+        eta=3,
+        asynchronous=True,
+        random_state=42,
+        scoring=None,
+        test_size=0.15,
+    ):
+        self.model = model
+        self.params = params
+        self.max_iter = max_iter
+        self.eta = eta
+        self.test_size = test_size
+        self.random_state = random_state
+        self.asynchronous = asynchronous
```

Comment:
Out of curiosity, you seem to prefer the asynchronous case.  Are there situations when it isn't better?

Reply:
Yes. The synchronous version is appropriate when there are few workers. I've defaulted to the asynchronous case because we're using dask.distributed. Future work might be to make a smart choice here between the sync and async versions.

This is because the synchronous Hyperband has to wait for other models to complete before calling `partial_fit` again. This is a problem if there are many workers: workers will be sitting idle.

I have some images of this waiting in the sync version and not waiting in the async version in https://github.com/dask/dask-ml/pull/221#issuecomment-400869493